File: src/android/FirebasePluginMessageReceiverManager.java
package org.apache.cordova.firebasex;

import android.os.Bundle;

import com.google.firebase.messaging.RemoteMessage;

import java.util.ArrayList;
import java.util.List;

/**
 * Singleton manager that tracks all registered {@link FirebasePluginMessageReceiver} instances
 * and dispatches incoming FCM messages and notification bundles to them.
 *
 * <p>When a message or bundle arrives, each registered receiver is invoked in registration order.
 * If any receiver returns {@code true}, the manager reports the message as handled, but all
 * receivers are still invoked (i.e., handling does not short-circuit).
 *
 * @see FirebasePluginMessageReceiver
 */
public class FirebasePluginMessageReceiverManager {

    /** List of registered message receivers, in registration order. */
    private static List<FirebasePluginMessageReceiver> receivers = new ArrayList<FirebasePluginMessageReceiver>();

    /**
     * Registers a message receiver. Typically called automatically from
     * {@link FirebasePluginMessageReceiver#FirebasePluginMessageReceiver()}.
     *
     * @param receiver the receiver to register
     */
    public static void register(FirebasePluginMessageReceiver receiver) {
        receivers.add(receiver);
    }

    /**
     * Dispatches an incoming FCM message to all registered receivers.
     *
     * @param remoteMessage the incoming FCM {@link RemoteMessage}
     * @return {@code true} if at least one receiver handled the message
     */
    public static boolean onMessageReceived(RemoteMessage remoteMessage) {
        boolean handled = false;
        for (FirebasePluginMessageReceiver receiver : receivers) {
            boolean wasHandled = receiver.onMessageReceived(remoteMessage);
            if (wasHandled) {
                handled = true;
            }
        }
        return handled;
    }

    /**
     * Dispatches a notification bundle to all registered receivers.
     *
     * @param bundle the notification data bundle
     * @return {@code true} if at least one receiver handled the bundle
     */
    public static boolean sendMessage(Bundle bundle) {
        boolean handled = false;
        for (FirebasePluginMessageReceiver receiver : receivers) {
            boolean wasHandled = receiver.sendMessage(bundle);
            if (wasHandled) {
                handled = true;
            }
        }
        return handled;
    }
}

File: src/ios/FirebasePluginMessageReceiverManager.m
/**
 * @file FirebasePluginMessageReceiverManager.m
 * @brief Implementation of the message receiver manager for iOS.
 *
 * Maintains a static list of registered @c FirebasePluginMessageReceiver instances
 * and dispatches notifications to each one when called by the messaging plugin.
 */

#import "FirebasePluginMessageReceiverManager.h"

@implementation FirebasePluginMessageReceiverManager

/** Lazily-initialised list of registered message receiver instances. */
static NSMutableArray<FirebasePluginMessageReceiver *> *receivers;

/**
 * Registers a message receiver, lazily creating the receivers array if needed.
 *
 * @param receiver The receiver to register.
 */
+ (void)registerMessageReceiver:(FirebasePluginMessageReceiver *)receiver {
    if (receivers == nil) {
        receivers = [[NSMutableArray alloc] init];
    }
    [receivers addObject:receiver];
}

/**
 * Dispatches a notification to all registered receivers.
 *
 * If no receivers are registered (array is @c nil), the call is a no-op.
 *
 * @param notification The notification payload dictionary.
 */
+ (void)sendNotification:(NSDictionary *)notification {
    if (receivers == nil) return;
    for (FirebasePluginMessageReceiver *receiver in receivers) {
        [receiver sendNotification:notification];
    }
}

@end
